src/parse.rs

```diff
@@ -60,52 +106,27 @@ impl Request {
         }
 
         match (query, fragment) {
-            (None, None) => (SmolStr::new(ident), SmolStr::default(), SmolStr::default()),
-            (None, Some(index)) => (
-                SmolStr::new(&ident[0..index]),
-                SmolStr::default(),
-                SmolStr::new(&ident[index..]),
-            ),
-            (Some(index), None) => (
-                SmolStr::new(&ident[0..index]),
-                SmolStr::new(&ident[index..]),
-                SmolStr::default(),
-            ),
+            (None, None) => (ident.into(), None, None),
+            (None, Some(j)) => (ident[0..j].into(), None, Some(ident[j..].into())),
+            (Some(i), None) => (ident[0..i].into(), Some(ident[i..].into()), None),
             (Some(i), Some(j)) => (
-                SmolStr::new(&ident[0..i]),
-                SmolStr::new(&ident[i..j]),
-                SmolStr::new(&ident[j..]),
+                ident[0..i].into(),
+                Some(ident[i..j].into()),
+                Some(ident[j..].into()),
             ),
         }
     }
 
-    pub(crate) fn with_target(self, target: &str) -> Self {
-        let is_directory = Self::is_directory(target);
-        Self {
-            kind: Resolver::get_target_kind(target),
-            target: target.into(),
-            is_directory,
-            ..self
-        }
-    }
-
     #[inline]
-    fn is_directory(target: &str) -> bool {
+    fn is_target_directory(target: &str) -> bool {
         target.ends_with('/')
     }
 }
 
 impl Resolver {
-    pub(crate) fn parse(&self, request: &str) -> Request {
-        let (target, query, fragment) = Request::parse_identifier(request);
-        let is_directory = Request::is_directory(&target);
-        Request {
-            kind: Self::get_target_kind(&target),
-            target,
-            query,
-            fragment,
-            is_directory,
-        }
+    #[must_use]
+    pub(crate) fn parse(request: &str) -> Request {
+        Request::from_request(request)
```

Comment:
This is a bad api (in Rust terms), I'll remove it in the next few PRs.

Reply:
maybe we need `trait From` ?